tftpd/tftpd.c

```diff
@@ -1631,16 +1651,31 @@ static void tftp_sendfile(const struct formats *pf, struct tftphdr *oap, int oac
         }
         dp->th_opcode = htons((u_short) DATA);
         dp->th_block = htons((u_short) block);
-        timeout = rexmtval;
-        (void)sigsetjmp(timeoutbuf, 1);
+
+        offset -= size;
+        window++;
+
+        if (sigsetjmp(timeoutbuf, 1) != 0) {
+            if (lseek(fileno(file), offset, SEEK_CUR) < 0) {
+                syslog(LOG_WARNING, "tftpd: read(ack): %m");
+                goto abort;
+            }
+            block -= (window - 1);
+            window = offset = 0;
+            dp = r_init();
+            continue;
+        }
 
         r_timeout = timeout;
         if (send(peer, dp, size + 4, 0) != size + 4) {
             syslog(LOG_WARNING, "tftpd: write: %m");
             goto abort;
         }
-        read_ahead(file, pf->f_convert);
-        for (;;) {
+
+        // TODO: make this useful
+        // read_ahead(file, pf->f_convert);
+
+        while (window == windowsize || size != segsize) {
```

Comment:
This is ugly. Basically, this condition is related to the previous code. I would do something in the term of
for(window=0;window<windowsize;window++) {
send data
}
for(;;;) {
wait and parse the reply
}